### PR TITLE
GUPSGen

### PR DESCRIPTION
**Pull request by mahyarsamani**

### CODE REVIEW

**Comment by mahyarsamani** [comment]
I'm still debugging this code and there are stages that are not yet implemented, but I guess you guys can already review the code.

**Comment by mahyarsamani** [comment]
Sorry some of the changes from gerrit/stable have been moved here as my change, for the time being just look at src/cpu/testers/traffic_gen/gups_gen.cc and src/cpu/testers/traffic_gen/gups_gen.hh

File: src/cpu/testers/traffic_gen/gups_gen.cc [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    port(name() + ".port", this),
+    startAddr(params.start_addr),
+    memSize(params.mem_size),
+    blockSize(params.block_size)
+{}
+
+Port&
+GUPSGen::getPort(const std::string &if_name, PortID idx)
+{
+    if (if_name != "port"){
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.cc:27]: style: Space after `)`

Please fix everywhere.

I've mentioned this before, but don't wait on style stuff to the end. Just do it right as you write the code.

File: src/cpu/testers/traffic_gen/gups_gen.cc [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+    // std::fill_n(pkt_data, req->getSize(), (uint8_t)requestorId);
+    std::memcpy(pkt_data, data, req->getSize());
+
+    return pkt;
+}
+
+
+uint64_t
+GUPSGen::randomNumberGenerator(int64_t n)
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.cc:174]: Is this defined in GUPS? Why not use the random functions from gem5/base?

File: src/cpu/testers/traffic_gen/gups_gen.cc [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+    PacketPtr pkt = requestPool.front();
+
+    exitTimes[pkt] = curTick();
+    port.sendTimingPacket(pkt);
+
+    requestPool.pop();
+
+    if (!nextSendEvent.scheduled() && !requestPool.empty())
+    {
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.cc:126]: Previous line...

File: src/cpu/testers/traffic_gen/gups_gen.hh [hunk 4]
```diff
@@ -0,0 +1,147 @@
+#ifndef __MEM_GUPS_GEN_HH__
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.hh:1]: LICENSE

File: src/cpu/testers/traffic_gen/gups_gen.hh [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+#include "base/statistics.hh"
+#include "mem/port.hh"
+#include "params/GUPSGen.hh"
+#include "sim/clocked_object.hh"
+#include "sim/system.hh"
+
+class GUPSGen : public ClockedObject {
+
+    private:
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.hh:16]: Two spaces, not 4

File: src/cpu/testers/traffic_gen/gups_gen.hh [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        System *const system;
+
+        const RequestorID requestorId;
+
+        class MemSidePort : public RequestPort
+        {
+            private:
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.hh:23]: two spaces

File: src/cpu/testers/traffic_gen/gups_gen.hh [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+        virtual void init() override;
+
+        bool handleResponse(PortID memPortId, PacketPtr pkt);
+
+
+        PacketPtr getReadPacket(Addr);
+        PacketPtr getWritePacket(Addr, uint8_t*);
+
+        void generateNextBatch();
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/gups_gen.hh:91]: Some comments would be helpful

**Review by powerjg — COMMENTED** [review]
A few comments on what you have so far

**Pull request closed without merging**